File: examples/Ruby/EX7/telephone.rb
#!/usr/bin/env ruby
# -*- tab-width: 4; -*-

#
# The contents of this file are subject to the Mozilla Public
# License Version 1.1 (the "License"); you may not use this file
# except in compliance with the License. You may obtain a copy of
# the License at http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS
# IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
# implied. See the License for the specific language governing
# rights and limitations under the License.
#
# The Original Code is State Machine Compiler (SMC).
#
# The Initial Developer of the Original Code is Charles W. Rapp.
# Portions created by Charles W. Rapp are
# Copyright (C) 2000 - 2003 Charles W. Rapp.
# All Rights Reserved.
#
# Contributor(s):
#       Port to Ruby by Francois Perrad, francois.perrad@gadz.org
#
# Name
#  telephone.rb
#
# Description
#  A simulation of an old fashioned touch-tone telephone.
#
# RCS ID
# $Id: telephone.rb,v 1.2 2009/04/11 14:29:14 cwrapp Exp $
#
# CHANGE LOG
# $Log: telephone.rb,v $
# Revision 1.2  2009/04/11 14:29:14  cwrapp
# Added called to enterStartState.
#
# Revision 1.1  2005/06/16 17:52:04  fperrad
# Added Ruby examples 1 - 4 and 7.
#
#

require 'time'
require 'tk'

require 'Telephone_sm'

class Telephone

	LONG_DISTANCE = 1
	LOCAL = 2
	EMERGENCY = 3

	NYC_TEMP = 4
	TIME = 5
	DEPOSIT_MONEY = 6
	LINE_BUSY = 7
	INVALID_NUMBER = 8

	SEC_PER_MINUTE = 60

	def initialize(root)
		@_root = root
		@_areaCode = ""
		@_exchange = ""
		@_local = ""
		@_display = ""
		@_receiverButton = nil
		@_timerMap = {
				'ClockTimer'=> nil,
				'OffHookTimer'=> nil,
				'LoopTimer'=> nil,
				'RingTimer'=> nil,
		}
		@_timerAudioID = nil

		_loadUI

		# Create the state machine to drive this object.
		@_fsm = Telephone_sm::new(self)

		# DEBUG
		#@_fsm.setDebugFlag(true)
	end


	# Create the user interface but don't display it yet.
	def _loadUI()
		@_root.title("Telephone demo")

		frameDisplay = TkFrame.new(@_root)
		frameDisplay.pack(
				'side' => 'top',
				'fill' => 'both'
		)
		# Create the read-only phone number display.
		@_numberDisplay = TkLabel.new(frameDisplay,
			'width' => 30,
			'bg' => 'white',
			'relief' => 'sunken',
			'padx' => 5,
			'pady' => 5
		)
		@_numberDisplay.pack(
			'side' => 'top',
			'padx' => 5,
			'pady' => 5
		)

		frameHook = TkFrame.new(@_root)
		frameHook.pack(
			'side' => 'top',
			'expand' => 1
		)
		# Create the off-hook/on-hook button.
		@_receiverButton = TkButton.new(frameHook,
			'text' =>  "Pick up receiver",
			'state' => 'normal',
			'command' => proc {
				text = @_receiverButton.cget('text')
				if text == "Pick up receiver" then
					@_fsm.OffHook
				elsif text == "Put down receiver" then
					@_fsm.OnHook
				else
					$stderr.printf("Unknown receiver (%s).\n", text)
				end
			},
			'padx' => 10,
			'pady' => 5,
			'bd'	=>  3
		)
		@_receiverButton.pack(
				'side' => 'top',
				'padx' => 5,
				'pady' => 5
		)

		frameDial = TkFrame.new(@_root)
		frameDial.pack(
			'side' => 'top',
			'fill' => 'both',
			'padx' => 5,
			'pady' => 5
		)
		# Create the dialing buttons.
		i = 0
		"123456789*0#".split(//).each do |digit|
			b = TkButton.new(frameDial,
				'text' => digit,
				'height' => 2,
				'width' => 3,
				'command' => proc { @_fsm.Digit(digit) }
			)
			b.grid(
				'row' => i/3,
				'column' => i%3,
				'padx' => 1,
				'pady' => 1
			)
			i += 1
		end

		frameStatus = TkFrame.new(@_root)
		frameStatus.pack(
			'side' => 'top',
			'fill' => 'both'
		)
		@_soundDisplay = TkLabel.new(frameStatus,
			'relief' => 'groove'
		)
		@_soundDisplay.pack(
			'side' => 'top',
			'fill' => 'x',
			'expand' => 1,
			'anchor' => 's'
		)

		# Cntl-C stops the demo as well.
		@_root.bind('Control-c') { exit }
	end

    def startFSM()
        @_fsm.enterStartState
    end

	#-----------------------------------------------------------
	# State Machine Actions.
	#

	# Return the current area code.
	def getAreaCode()
		return @_areaCode
	end

	# Return the exchange.
	def getExchange()
		return @_exchange
	end

	# Return the local number.
	def getLocal()
		return @_local
	end

	def routeCall(callType, areaCode, exchange, local)
		if callType == EMERGENCY then
			route = EMERGENCY
		elsif (callType == LONG_DISTANCE and
			areaCode == "1212" and
			exchange == "555" and
			local == "1234") then
			route = NYC_TEMP
		elsif exchange == "555" then
			if local == "1212" then
				route = TIME
			else
				route = LINE_BUSY
			end
		elsif callType == LOCAL then
			route = DEPOSIT_MONEY
		else
			route = INVALID_NUMBER
		end

		# Call routing needs to be done asynchronouzly in order to
		# avoid issuing a transition within a transition.
		Tk.after(50, proc { _callRoute(route) } )
	end

	def startTimer(name, delay)
		if name == "ClockTimer" then
			@_timerMap[name] = TkAfter.new(delay, 1, proc { @_fsm.ClockTimer } )
		elsif name == "OffHookTimer" then
   			@_timerMap[name] = TkAfter.new(delay, 1, proc { @_fsm.OffHookTimer } )
		elsif name == "LoopTimer" then
			@_timerMap[name] = TkAfter.new(delay, 1, proc { @_fsm.LoopTimer } )
		elsif name == "RingTimer" then
			@_timerMap[name] = TkAfter.new(delay, 1, proc { @_fsm.RingTimer } )
		end
		@_timerMap[name].start
	end

	def resetTimer(name)
		unless @_timerMap[name].nil? then
			@_timerMap[name].restart
		end
	end

	def stopTimer(name)
		unless @_timerMap[name].nil? then
			@_timerMap[name].cancel
			@_timerMap[name] = nil
		end
	end

	def play(name, delay)
		unless @_timerAudioID.nil? then
			@_timerAudioID.cancel
			@_timerAudioID = nil
		end
		@_soundDisplay.configure('text' => name)
		@_timerAudioID = TkAfter.new(delay, 1, proc {
			@_soundDisplay.configure('text' => "")
			@_timerAudioID = nil
		} )
		@_timerAudioID.start
	end

	def playTT(name)
		play(name, 400)
	end

	def loop(name)
		unless @_timerAudioID.nil? then
			@_timerAudioID.cancel
			@_timerAudioID = nil
		end
		@_soundDisplay.configure('text' => name + " ...")
	end

	def stopLoop(name)
		@_soundDisplay.configure('text' => "")
	end

	def stopPlayback()
		unless @_timerAudioID.nil? then
			@_timerAudioID.cancel
			@_timerAudioID = nil
		end
		@_soundDisplay.configure('text' => "")
	end

	def playEmergency()
		play("911", 5000)
	end

	def playNYCTemp()
		play("NYC_temp", 2000)
	end

	def playDepositMoney()
		play("50_cents_please", 2000)
	end

	def playTime()
		play("the_time_is ???", 2000)
	end

	def playInvalidNumber()
		play("you_dialed ### could_not_be_completed", 2000)
	end

	def getType()
		return @_callType
	end

	def setType(type)
		@_callType = type
	end

	def saveAreaCode(n)
		@_areaCode += n
		addDisplay(n)
	end

	def saveExchange(n)
		@_exchange += n
		addDisplay(n)
	end

	def saveLocal(n)
		@_local += n
		addDisplay(n)
	end

	def addDisplay(character)
		@_display += character
		@_numberDisplay.configure('text' => @_display)
	end

	def clearDisplay()
		#Clear the internal data store.
		@_display = ""
		@_areaCode = ""
		@_local = ""
		@_exchange = ""

		# Put up the current time and date on the display.
		@_numberDisplay.configure('text' => "")
	end

	def startClockTimer()
		currentTime = Time::now.to_i
		timeRemaining = SEC_PER_MINUTE - (currentTime % SEC_PER_MINUTE)

		# Figure out how long until the top of the minute
		# and set the timer for that amount.
		startTimer("ClockTimer", timeRemaining * 1000)
	end

	def updateClock()
		text = Time::now.localtime.strftime("%H:%M  %b %d, %Y")
		@_numberDisplay.configure('text' => text)
	end

	def setReceiver(text)
		@_receiverButton.configure('text' => text)
	end

	def _callRoute(route)
		if route == EMERGENCY then
			@_fsm.Emergency
		elsif route == NYC_TEMP then
			@_fsm.NYCTemp
		elsif route == TIME then
			@_fsm.Time
		elsif route == DEPOSIT_MONEY then
			@_fsm.DepositMoney
		elsif route == LINE_BUSY then
			@_fsm.LineBusy
		elsif route == INVALID_NUMBER then
			@_fsm.InvalidNumber
		end
	end

end

# Display the "telephone" user interface and run until
# the user quits the window.
root = TkRoot.new
tel = Telephone.new(root)
tel.startFSM
Tk.mainloop

File: examples/Python/EX7/telephone.py
#!/usr/bin/env python
# -*- tab-width: 4; -*-

#
# The contents of this file are subject to the Mozilla Public
# License Version 1.1 (the "License"); you may not use this file
# except in compliance with the License. You may obtain a copy of
# the License at http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS
# IS" basis, WITHOUT WARRANTY OF ANY KIND, either express or
# implied. See the License for the specific language governing
# rights and limitations under the License.
#
# The Original Code is State Machine Compiler (SMC).
#
# The Initial Developer of the Original Code is Charles W. Rapp.
# Portions created by Charles W. Rapp are
# Copyright (C) 2000 - 2003 Charles W. Rapp.
# All Rights Reserved.
#
# Contributor(s):
#       Port to Python by Francois Perrad, francois.perrad@gadz.org
#
# Name
#  telephone.py
#
# Description
#  A simulation of an old fashioned touch-tone telephone.
#
# RCS ID
# $Id: telephone.py,v 1.3 2009/04/19 14:39:48 cwrapp Exp $
#
# CHANGE LOG
# $Log: telephone.py,v $
# Revision 1.3  2009/04/19 14:39:48  cwrapp
# Added call to enterStartState before issuing first FSM transition.
#
# Revision 1.2  2005/06/08 11:09:13  cwrapp
# + Updated Python code generator to place "pass" in methods with empty
#   bodies.
# + Corrected FSM errors in Python example 7.
# + Removed unnecessary includes from C++ examples.
# + Corrected errors in top-level makefile's distribution build.
#
# Revision 1.1  2005/05/28 17:48:30  cwrapp
# Added Python examples 1 - 4 and 7.
#
#

import sys, time
from Tkinter import *

import Telephone_sm

class Telephone:

	LONG_DISTANCE = 1
	LOCAL = 2
	EMERGENCY = 3

	NYC_TEMP = 4
	TIME = 5
	DEPOSIT_MONEY = 6
	LINE_BUSY = 7
	INVALID_NUMBER = 8

	SEC_PER_MINUTE = 60

	def __init__(self, root):
		self._root = root
		self._areaCode = ""
		self._exchange = ""
		self._local = ""
		self._display = ""
		self._receiverButton = None
		self._timerMap = {
				'ClockTimer': -1,
				'OffHookTimer': -1,
				'LoopTimer': -1,
				'RingTimer': -1,
		}
		self._timerAudioID = -1

		self._loadUI()

		# Create the state machine to drive this object.
		self._fsm = Telephone_sm.Telephone_sm(self)

		# DEBUG
		#self._fsm.setDebugFlag(True)


	# Create the user interface but don't display it yet.
	def _loadUI(self):
		self._root.title("Telephone demo")

		frameDisplay = Frame(self._root)
		frameDisplay.pack(
				side=TOP,
				fill=BOTH,
		)
		# Create the read-only phone number display.
		self._numberDisplay = Label(frameDisplay,
				width=30,
				bg='white',
				relief=SUNKEN,
				padx=5,
				pady=5,
		)
		self._numberDisplay.pack(
				side=TOP,
				padx=5,
				pady=5,
		)

		frameHook = Frame(self._root)
		frameHook.pack(
				side=TOP,
				expand=1,
		)
		# Create the off-hook/on-hook button.
		def onReceiver():
			text = self._receiverButton.cget('text')
			if text == "Pick up receiver":
				self._fsm.OffHook()
			elif text == "Put down receiver":
				self._fsm.OnHook()
			else:
				sys.stderr.write("Unknown receiver (%s).\n" % text)

		self._receiverButton = Button(frameHook,
				text="Pick up receiver",
				state=NORMAL,
				command=onReceiver,
				padx=10,
				pady=5,
				bd=3,
		)
		self._receiverButton.pack(
				side=TOP,
				padx= 5,
				pady= 5,
		)

		frameDial = Frame(self._root)
		frameDial.pack(
				side=TOP,
				fill=BOTH,
				padx=5,
				pady=5,
		)
		# Create the dialing buttons.
		i = 0
		for digit in ("123456789*0#"):
			b = Button(frameDial,
					text=digit,
					height=2,
					width=3,
					command=lambda digit=digit: self._fsm.Digit(digit),
			)
			b.grid(
					row=i/3,
					column=i%3,
					padx=1,
					pady=1,
			)
			i += 1

		frameStatus = Frame(self._root)
		frameStatus.pack(
				side=TOP,
				fill=BOTH,
		)
		self._soundDisplay = Label(frameStatus,
				relief=GROOVE,
		)
		self._soundDisplay.pack(
				side=TOP,
				fill='x',
				expand=1,
				anchor='s',
		)

		# Cntl-C stops the demo as well.
		self._root.bind('<Control-c>', lambda x: sys.exit(0))

	def Start(self):
		self._fsm.enterStartState()

	#-----------------------------------------------------------
	# State Machine Actions.
	#

	# Return the current area code.
	def getAreaCode(self):
		return self._areaCode

	# Return the exchange.
	def getExchange(self):
		return self._exchange

	# Return the local number.
	def getLocal(self):
		return self._local

	def routeCall(self, callType, areaCode, exchange, local):
		if callType == self.EMERGENCY:
			route = self.EMERGENCY
		elif (callType == self.LONG_DISTANCE
			and areaCode == "1212"
			and exchange == "555"
			and local == "1234"):
			route = self.NYC_TEMP
		elif exchange == "555":
			if local == "1212":
				route = self.TIME
			else:
				route = self.LINE_BUSY
		elif callType == self.LOCAL:
			route = self.DEPOSIT_MONEY
		else:
			route = self.INVALID_NUMBER

		# Call routing needs to be done asynchronouzly in order to
		# avoid issuing a transition within a transition.
		self._root.after(50, lambda: self._callRoute(route))

	def startTimer(self, name, delay):
		if name == "ClockTimer":
			self._timerMap[name] = self._root.after(delay, lambda: self._fsm.ClockTimer())
		elif name == "OffHookTimer":
			self._timerMap[name] = self._root.after(delay, lambda: self._fsm.OffHookTimer())
		elif name == "LoopTimer":
			self._timerMap[name] = self._root.after(delay, lambda: self._fsm.LoopTimer())
		elif name == "RingTimer":
			self._timerMap[name] = self._root.after(delay, lambda: self._fsm.RingTimer())

	def resetTimer(self, name, delay):
		if self._timerMap[name] >= 0:
			self._root.after_cancel(self._timerMap[name])
			self._timerMap[name] = -1;
			self.startTimer(name, delay)

	def stopTimer(self, name):
		if self._timerMap[name] >= 0:
			self._root.after_cancel(self._timerMap[name])
			self._timerMap[name] = -1;

	def play(self, name, delay):
		def endPlay():
			self._soundDisplay.configure(text="")
			self._timerAudioID = -1

		if self._timerAudioID >= 0:
			self._root.after_cancel(self._timerAudioID)
			self._timerAudioID = -1
		self._soundDisplay.configure(text=name)
		self._timerAudioID = self._root.after(delay, endPlay)

	def playTT(self, name):
		self.play(name, 400)

	def loop(self, name):
		if self._timerAudioID >= 0:
			self._root.after_cancel(self._timerAudioID)
			self._timerAudioID = -1
		self._soundDisplay.configure(text=name + " ...")

	def stopLoop(self, name):
		self._soundDisplay.configure(text="")

	def stopPlayback(self):
		if self._timerAudioID >= 0:
			self._root.after_cancel(self._timerAudioID)
			self._timerAudioID = -1
		self._soundDisplay.configure(text="")

	def playEmergency(self):
		self.play("911", 5000)

	def playNYCTemp(self):
		self.play("NYC_temp", 2000)

	def playDepositMoney(self):
		self.play("50_cents_please", 2000)

	def playTime(self):
		self.play("the_time_is ???", 2000)

	def playInvalidNumber(self):
		self.play("you_dialed ### could_not_be_completed", 2000)

	def getType(self):
		return self._callType

	def setType(self, type):
		self._callType = type

	def saveAreaCode(self, n):
		self._areaCode += n
		self.addDisplay(n)

	def saveExchange(self, n):
		self._exchange += n
		self.addDisplay(n)

	def saveLocal(self, n):
		self._local += n
		self.addDisplay(n)

	def addDisplay(self, character):
		self._display += character
		self._numberDisplay.configure(text=self._display)

	def clearDisplay(self):
		#Clear the internal data store.
		self._display = ""
		self._areaCode = ""
		self._local = ""
		self._exchange = ""

		# Put up the current time and date on the display.
		self._numberDisplay.configure(text="")

	def startClockTimer(self):
		currentTime = int(time.time())
		timeRemaining = self.SEC_PER_MINUTE - (currentTime % self.SEC_PER_MINUTE)

		# Figure out how long until the top of the minute
		# and set the timer for that amount.
		self.startTimer("ClockTimer", timeRemaining * 1000)

	def updateClock(self):
		text = time.strftime("%H:%M  %b %d, %Y", time.localtime(time.time()))
		self._numberDisplay.configure(text=text)

	def setReceiver(self, text):
		self._receiverButton.configure(text=text)

	def _callRoute(self, route):
		if route == self.EMERGENCY:
			self._fsm.Emergency()
		elif route == self.NYC_TEMP:
			self._fsm.NYCTemp()
		elif route == self.TIME:
			self._fsm.Time()
		elif route == self.DEPOSIT_MONEY:
			self._fsm.DepositMoney()
		elif route == self.LINE_BUSY:
			self._fsm.LineBusy()
		elif route == self.INVALID_NUMBER:
			self._fsm.InvalidNumber()

# Display the "telephone" user interface and run until
# the user quits the window.
root = Tk()
tel = Telephone(root)
tel.Start()
root.mainloop()
